File: src/main/java/com/baidu/algorithm/math/_233_Number_of_Digit_One.py
class Solution(object):
    def countDigitOne(self, n):
        """
        :type n: int
        :rtype: int
        """
        res, m = 0, 1
        while n >= m:
            a, b = divmod(n, m)

            c = a % 10
            a /= 10

            if c == 0:
                res += a * m
            elif c == 1:
                res += a * m + b + 1
            else:
                res += (a + 1) * m

            m *= 10

        return res

File: src/main/java/com/baidu/algorithm/math/_233_Number_of_Digit_One.java
/*
 * Copyright (C) 2018 Baidu, Inc. All Rights Reserved.
 */
package com.baidu.algorithm.math;

/**
 * _233_Number_of_Digit_One
 *
 * @author xuhaoran01
 */
public class _233_Number_of_Digit_One {

    public int countDigitOne(int n) {
        long res = 0, factor = 1;
        while (n >= factor) {
            long high = n / (factor * 10);
            long cur = (n / factor) % 10;
            long low = n % factor;

            if (cur == 0) {
                res += high * factor;
            } else if (cur == 1) {
                res += high * factor + low + 1;
            } else {
                res += (high + 1) * factor;
            }

            factor *= 10;
        }

        return (int) res;
    }
}
